# Day 93 - LocalOutlierFactor & IsolationForest

1. Load blobs.csv file into the df DataFrame. The file contains two variables x1 and x2. <br>
Use the LocalOutlierFactor class from the scikit-learn to analyze outliers in the given dataset. Set the following arguments: <br>
<br>
n_neighbors=20 <br>
<br>
As a reminder, 1 is the normal element, -1 is the outlier. Assign a new column to the df DataFrame named 'lof' that stores information whether a given sample is a normal or an outlier. In response, print the first ten rows of the df DataFrame to the console.

In [1]:
import numpy as np
import pandas as pd
from sklearn.neighbors import LocalOutlierFactor
 
np.random.seed(42)
df = pd.read_csv('blobs.csv')
data = df.values
lof = LocalOutlierFactor(n_neighbors=20)
y_pred = lof.fit_predict(data)
 
df['lof'] = y_pred
print(df.head(10))

          x2         x1  lof
0  -7.798349  -8.579798    1
1  -8.600454  -7.649221    1
2  -0.864108   6.572599    1
3   4.204516   4.170723    1
4 -10.955876  -8.896282    1
5   6.355198   1.653293    1
6  -2.162836   9.784921    1
7   5.465742   0.845721    1
8  -3.467546   8.642968    1
9  -4.891805  10.327393    1


2. Load blobs.csv file into the df DataFrame. The file contains two variables x1 and x2.  <br>
The LocalOutlierFactor class from the scikit-learn was used to analyze outliers from this dataset. <br>
As a reminder, 1 is the normal element, -1 is the outlier. A new column was assigned to the df object named lof, which stores the information whether a given sample is normal or not: <br>
<br>
          x2         x1  lof <br>
0  -7.798349  -8.579798    1 <br>
1  -8.600454  -7.649221    1 <br>
2  -0.864108   6.572599    1 <br>
3   4.204516   4.170723    1 <br>
4 -10.955876  -8.896282    1 <br>
5   6.355198   1.653293    1 <br>
6  -2.162836   9.784921    1 <br>
7   5.465742   0.845721    1 <br>
8  -3.467546   8.642968    1 <br>
9  -4.891805  10.327393    1 <br>
<br>
Find the number of outliers in this dataset. In response, print the result to the console as shown below. v
<br>
Expected Result: <br>
<br>
 1    290 <br>
-1     10 <br>
Name: lof, dtype: int64 

In [2]:
import numpy as np
import pandas as pd
from sklearn.neighbors import LocalOutlierFactor
 
np.random.seed(42)
df = pd.read_csv('blobs.csv')
data = df.values
lof = LocalOutlierFactor(n_neighbors=30)
y_pred = lof.fit_predict(data)
 
df['lof'] = y_pred
print(df['lof'].value_counts())

 1    290
-1     10
Name: lof, dtype: int64


3. Load factory.csv file into the df DataFrame. This file contains two variables item_length and item_width. <br>
Using the IsolationForest class from the scikit-learn, analyze the outliers in this dataset. Set following arguments: <br>
<br>
n_estimators=100 <br>
contamination=0.05 <br>
random_state=42 <br>
<br>
As a reminder, 1 is the normal element, -1 is the outlier. Assign a new column to the df object named 'outlier_flag' that stores information whether the sample is normal or outlier. In response, print the first ten rows of the df object to the console.

In [3]:
import numpy as np
import pandas as pd
from sklearn.ensemble import IsolationForest
  
np.random.seed(42)
df = pd.read_csv('factory.csv')
outlier = IsolationForest(
    n_estimators=100, contamination=0.05, random_state=42
)
outlier.fit(df)
df['outlier_flag'] = outlier.predict(df)
print(df.head(10))

e:\python\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but IsolationForest was fitted with feature names
  warnings.warn(


   item_length  item_width  outlier_flag
0     6.206380    5.618286             1
1     5.756649    5.921924             1
2     5.576329    5.792401             1
3     5.304950    7.235000             1
4     6.764064    5.965792             1
5     5.668838    5.708257             1
6     6.241278    6.533104             1
7     5.951632    6.503750             1
8     7.192132    6.249855             1
9     5.999551    7.660598             1


4. The factory.csv file was loaded into the df DataFrame. This file contains two variables item_length and item_width. <br>
The IsolationForest class from the scikit-learn was used to detect outliers from the given dataset. <br>
As a reminder, 1 is the normal element, -1 is the outlier. A new column was assigned to the df object named outlier_flag, which stores the information whether a given sample is normal or not: <br>
<br>
   item_length  item_width  outlier_flag <br>
0     6.206380    5.618286             1 <br>
1     5.756649    5.921924             1 <br>
2     5.576329    5.792401             1 <br>
3     5.304950    7.235000             1 <br>
4     6.764064    5.965792             1 <br>
5     5.668838    5.708257             1 <br>
6     6.241278    6.533104             1 <br>
7     5.951632    6.503750             1 <br>
8     7.192132    6.249855             1 <br>
9     5.999551    7.660598             1 <br>
Find the number of outliers in this dataset and print the result to the console as shown below.

In [4]:
import numpy as np
import pandas as pd
from sklearn.ensemble import IsolationForest
 
np.random.seed(42)
df = pd.read_csv('factory.csv')
outlier = IsolationForest(
    n_estimators=100, contamination=0.05, random_state=42
)
outlier.fit(df)
df['outlier_flag'] = outlier.predict(df)
print(df['outlier_flag'].value_counts())

 1    380
-1     20
Name: outlier_flag, dtype: int64


e:\python\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but IsolationForest was fitted with feature names
  warnings.warn(
